source/04_best_practice/buffer.rst

```diff
@@ -0,0 +1,172 @@
+For example, in chess, Go, or card games where players are only rewarded at the end of the game, algorithms solving such tasks often want to process the entire game, and some algorithms like Hindsight Experience Replay (HER) need to sample complete episodes and process them in episodic units.
+In this case, the user can use group sampling to achieve this goal.
+
+- **Custom implementation via atomic operations**
+
+  When storing samples, you can add "episode" information to the meta to specify the episode to which the sample belongs, and when sampling, you can set groupby="episode" to enable group sampling with the episode keyword. \ **Sampling in groups can seriously increase the sampling time**\.
+
+  .. code-block:: python
+
+      buffer = DequeBuffer(size=10)
+
+      # When storing data, the user needs to add grouping information to the meta, e.g., "episode" as the grouping keyword, and the corresponding value is the specific group
+      buffer.push("a", {"episode": 1})
+      buffer.push("b", {"episode": 2})
+      buffer.push("c", {"episode": 2})
+
+      # Grouping according to the keyword "episode" requires that the number of different groups in the buffer is not less than the number of samples.
+      grouped_data = buffer.sample(2, groupby="episode")
+
+- **Implementation through middleware**
+
+  In the R2D2 algorithm, for example, the training samples are passed through the LSTM network in episodes, where group sampling is required.
+  In DI-engine, each env. corresponds to a separate decision track, so using env. as a keyword for grouping can have the effect of distinguishing between episodes.
+  The relevant code example is as follows, the full version of the code can be seen at \ `./ding/example/r2d2.py <https://github.com/opendilab/DI-engine/blob/main/ding/example/r2d2.py>`_：
+
+  .. code-block:: python
+
+      buffer = DequeBuffer(size=10)
+
+      # Here 'env' is used as the keyword for grouping, so that samples with the same env_id will be classified into the same group when sampling.
+      task.use(data_pusher(cfg, buffer, group_by_env=True))
+
+
+**(Available options)**
+On top of group sampling, you can also use \ **group_sample middleware**\  to implement post-processing of samples, such as: choosing whether to disrupt data within the same group, and setting the maximum length of each group of data.
+
+.. code-block:: python
+    
+    from ding.data.buffer.middleware import group_sample
+
+    buffer = DequeBuffer(size=10)
+    # The maximum length of each group of data is set to 3, keeping the original order within the group
+    buffer.use(group_sample(size_in_group=3, ordered_in_group=True))
+    
+
+**Delete multiple use samples**
+
+By default, the samples in the buffer may be repeatedly collected multiple times. During the training process, if these duplicate samples are used repeatedly without control, it may lead to poor results of the model.
```

Comment:
By default in the Dequebuffer, samples may be collected repeatedly by multiple sample function calls. If it is not controlled, the training performance will decay since it fits partial samples too many times.